src/components/inputs/SearchSelect.jsx

```diff
@@ -0,0 +1,427 @@
+    this._openSuggestedGroup(option);
+  },
+
+  _clearSuggestionOption() {
+    this._setSuggestedOption(null);
+  },
+
+  //State Helpers: filteredOptions
+
+  _getOptionsIterator(filterValue) {
+    let options = this._getFilteredOptions(filterValue);
+    return _.flatten(_.map(options, (option) => option.isGroup ? option.options : option));
+  },
+
+  _getFilteredOptions(filterValue) {
+    let {options, filterOption, hideGroupsWithNoMatch} = this.props;
+
+    //Filter grouped options
+    options = _.map(options, (group) => {
+      if (!group.isGroup) {
+        return group;
+      }
+      return {
+        ...group,
+        options: _.filter(group.options, (option) => !filterValue || filterOption(filterValue, option, group)),
+      };
+    });
+
+    //Remove empty groups if hideGroupsWithNoMatch == true
+    options = _.filter(options, (option) => !(option.isGroup && option.options.length === 0 && hideGroupsWithNoMatch));
+
+    //Filter non grouped options
+    options = _.filter(options, (option) => option.isGroup || (!filterValue || filterOption(filterValue, option)));
+
+    return options;
+  },
+
+  //Element Helpers
+
+  _openList() {
+    this.setState({isFocused: true});
+  },
+
+  _closeList() {
+    this._blurFilterInput();
+    this.setState({isFocused: false});
+    this._closeAllGroups();
+  },
+
+  _focusFilterInput() {
+    this.refs.searchInput.getDOMNode().focus();
+  },
+
+  _blurFilterInput() {
+    this.refs.searchInput.getDOMNode().blur();
+  },
+
+  //Elements Listeners
+
+  _cancelBlurInterval() {
+    if (this._blurInterval) {
+      clearTimeout(this._blurInterval);
+      this._blurInterval = null;
+    }
+  },
+
+  _onFocus(e) {
+    this._cancelBlurInterval();
+    this._openList();
+  },
+
+  _onBlur(e) {
+    this._cancelBlurInterval();
+    this._blurInterval = setTimeout(() => {
+      this._closeList();
+    }, 400);
+  },
+
+  _onSelectValueClick(e) {
+    this._focusFilterInput();
+  },
+
+  _onFilterInputChange(e) {
+    let newValue = e.target.value;
+    let previousValue = this.state.filterValue;
+
+    this._removeSelection();
+    this._updateFilterValue(newValue);
+
+    if (newValue === '') {
+      this._closeAllGroups();
+    }
+
+    let wasEmpty = previousValue.length === 0 && newValue.length > 0;
+    if (wasEmpty) {
+      this._openAllGroups();
+    }
+  },
+
+  _onFilterInputBlur(e) {
+    this._cancelBlurInterval();
+    this._blurInterval = setTimeout(() => {
+      this._clearFilterValue();
+      this._closeList();
+    }, 100);
+  },
+
+  _onFilterInputKeyDown(e) {
+    switch (e.which) {
+    case KEY_CODES.TAB:
+    case KEY_CODES.ENTER:
+      this._selectOption(this.state.suggestedOption);
+      break;
+    case KEY_CODES.ARROW_UP:
+      this._suggestPreviousOption(this.state.filterValue, this.state.suggestedOption);
+      break;
+    case KEY_CODES.ARROW_DOWN:
+      this._suggestNextOption(this.state.filterValue, this.state.suggestedOption);
```

Comment:
This have been modified later, the reason I did that is that when I chained functions that updated state, state might not be updated from one to an other.